### PR TITLE
Support inverse gates

### PR DESCRIPTION
**Pull request by trbromley**

### CODE REVIEW

File: pennylane_lightning/src/Apply.hpp [hunk 1]
```diff
@@ -79,12 +81,15 @@ namespace Pennylane {
      * @param ops list of unique string names corresponding to gate types, in the order they should be applied
      * @param wires list of wires on which each gate acts
      * @param params list of parameters that defines the gate parameterisation
+     * @param inverse list of bools indicating whether a given gate or its inverse should be applied
+     * @param qubits total number of qubits
```
Comment (trbromley) [inline pennylane_lightning/src/Apply.hpp:85]:
```suggestion
     * @param qubits number of qubits
```

File: pennylane_lightning/lightning_qubit.py [hunk 2]
```diff
@@ -74,6 +74,30 @@ class LightningQubit(DefaultQubit):
...
+        "CNOT",
+        "SWAP",
+        "CZ",
+        "CRX",
+        "CRY",
+        "CRZ",
+        "CRot",
+        "Toffoli",
+        "CSWAP",
+    }
```
Comment (trbromley) [inline pennylane_lightning/lightning_qubit.py:99]: These are the gates that have a kernel supported. Ideally, we would not explicitly set `operations` and just let it inherit from `default.qubit`. We'd then have a check in `apply_lightning()` that maps the operations to `QubitUnitary` if the operation is not in `kernel_operations`. In such a way, we're supporting essentially all ops, even without a kernel.

File: tests/test_gates.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            -0.04067799 + 0.34681783j,
+            -0.45852968 - 0.03214391j,
+            -0.10528164 - 0.4431247j,
+            0.50251451 + 0.45476965j,
+        ],
+    ]
+)
+
+
+@pytest.mark.xfail(strict=True)  # needs support for QubitUnitary
```
Comment (trbromley) [inline tests/test_gates.py:113]: When we add support for arbitrary unitaries, this will be a nice test for gates without a pre-defined kernel.

**Comment by trbromley** [comment]
The performance comparison looks like no change vs `master` (as expected):
image

**Review by antalszava — COMMENTED** [review]
Thanks @trbromley! 💯 My main question is related to whether we'd like to have new gate tests here apart from what's there in the device test suite. Also, would we like to unit test the kernel applications?

**Review by thisac — APPROVED** [review]
Nice work @trbromley! 🥇 

I have the same thought as @antalszava whether we want to add unit tests for the `applyKernel` functions, although I wouldn't mind if we decided to add them in a separate PR.

File: pennylane_lightning/src/Gates.cpp [hunk 4]
```diff
@@ -89,7 +98,8 @@ const vector<CplxType> Pennylane::XGate::matrix{
     0, 1,
     1, 0 };
 
-void Pennylane::XGate::applyKernel(const StateVector& state, const std::vector<size_t>& indices, const std::vector<size_t>& externalIndices) {
+void Pennylane::XGate::applyKernel(const StateVector& state, const std::vector<size_t>& indices, const std::vector<size_t>& externalIndices, bool inverse) {
+    (void)inverse;  // gate is its own inverse
```
[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane_lightning/src/Gates.cpp:102]: Hmm, how come we didn't have such warning for `applyKernel`? 🤔 We had a thread with Thomas on this here: https://github.com/PennyLaneAI/pennylane-lightning/pull/85#discussion_r592480332.

File: pennylane_lightning/lightning_qubit.py [hunk 5]
```diff
@@ -133,10 +157,15 @@ def apply_lightning(self, state, operations):
...
+        op_inverse = [o.inverse for o in operations]
 
         state_vector = np.ravel(state)
-        apply(state_vector, op_names, op_wires, op_param, self.num_wires)
+        apply(state_vector, op_names, op_wires, op_param, op_inverse, self.num_wires)
         return np.reshape(state_vector, state.shape)
+
+    @staticmethod
+    def _remove_inverse_string(string):
+        return string.split(".")[0]
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane_lightning/lightning_qubit.py]: Great suggestion! Have done.

File: pennylane_lightning/lightning_qubit.py [hunk 6]
```diff
@@ -133,10 +157,15 @@ def apply_lightning(self, state, operations):
...
         op_param = [o.parameters for o in operations]
+        op_inverse = [o.inverse for o in operations]
 
         state_vector = np.ravel(state)
-        apply(state_vector, op_names, op_wires, op_param, self.num_wires)
+        apply(state_vector, op_names, op_wires, op_param, op_inverse, self.num_wires)
         return np.reshape(state_vector, state.shape)
+
+    @staticmethod
+    def _remove_inverse_string(string):
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane_lightning/lightning_qubit.py:169]: Done!

File: pennylane_lightning/src/Gates.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane_lightning/src/Gates.cpp:102]: Yes, I think it's motivated for two reasons:
- So we won't get the warnings: `pennylane_lightning/src/Gates.cpp:101:147: warning: unused parameter ‘inverse’ [-Wunused-parameter]`
- I like having something that explicitly says the gate is it's own inverse. In other words, we're saying that we know about the `inverse` argument and don't need to do anything in this case, rather than we just forgot about it.

It'll work either way, so if we're opposed to having this line then I can remove. I don't think this has any performance implications, seems like something more for the compiler.

File: pennylane_lightning/src/Gates.cpp [hunk 7]
```diff
@@ -174,10 +187,13 @@ const vector<CplxType> Pennylane::SGate::matrix{
     1, 0,
     0, IMAG };
 
-void Pennylane::SGate::applyKernel(const StateVector& state, const std::vector<size_t>& indices, const std::vector<size_t>& externalIndices) {
+void Pennylane::SGate::applyKernel(const StateVector& state, const std::vector<size_t>& indices, const std::vector<size_t>& externalIndices, bool inverse) {
+    CplxType shift = IMAG;
+    if (inverse == true) {shift *= -1;}
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane_lightning/src/Gates.cpp]: Nice, looks awesome, I'll make this change. One thing to note, it goes: `False ? True`

File: tests/test_gates.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+from pennylane_lightning import LightningQubit
+
+
+@pytest.fixture
+def op(op_name):
+    return getattr(qml, op_name)
+
+
+@pytest.mark.parametrize("op_name", LightningQubit.kernel_operations)
```
[the remark this answers is not in the document]
Comment (trbromley) [reply tests/test_gates.py]: That's a good point! From a quick look through those tests, I agree that they provide similar coverage to what's here. I'm generally happy to have some redundancy because:
- This is directly in `lightning.qubit` and it is easier for developers to quickly see this failing (often, at least for me, running the dev test suite is more of an afterthought)
- I didn't see this exact test in the suite, so maybe there is some edge case covered (not sure)!

Really, having it here explicitly makes me feel more confident that things are definitely working :laughing:

**Review by trbromley — COMMENTED** [review]
Thanks @antalszava @thisac!

**Comment by trbromley** [comment]
> Also, would we like to unit test the kernel applications?

It's a good question :thinking: My thinking is that we can move much more quickly when writing Python tests, so if we can access and test from the Python side then we should. I think the tests here manage to already provide a nice check of `applyKernel()`, even if they are not exactly unit tests.

File: tests/test_gates.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (antalszava) [reply tests/test_gates.py]: Agreed 🤔 Calling the device test suite can be overlooked. Could we
1) Call on the device test suite from the lightning tests (not quite sure about the best way to do this 🤔 )
2) extend `make test` with the device test command?

For 1), we could maybe call `pl-device-test` from one of the test files 🤔 

My worry is that we'd start off adding test cases here which will essentially diverge from the device test suite and we'll have to maintain both.

File: pennylane_lightning/src/Gates.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane_lightning/src/Gates.cpp:102]: Got it! It's odd that this warning is emitted 🤔 This is a personal opinion, but the statement looks confusing to me, as it's type casting a boolean to `void` without any effect. To me, having the parameter unnamed would make up for the unused parameter. The comment is indeed nice to indicate that the gate is its own inverse.

[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane_lightning/src/Gates.cpp]: Ok, updated to just have a commend and an unnamed param: fcd1c29a6755c013b916a651e2b724c834f0db9c

File: tests/test_gates.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply tests/test_gates.py]: Yes I agree maintainability is important! Because I'm paranoid about double making sure that things work correctly now for `lightning.qubit`, I still prefer to potentially have some repetition with a benefit of increased confidence. If down the line we see issues with these local tests, we could delete them (and potentially update the dev test suite accordingly).

**Review by antalszava — APPROVED** [review]
Looks good to me! 💯 Awesome that we can now also support inverses 😊
